### PR TITLE
docs(plugin): Add AppDynamics plugin documentation

### PR DESCRIPTION
**Pull request by hanshuebner**
### Summary

Add user facing documentation for the new AppDynamics plugin that will become part of Kong EE 3.1.0

### Reason

https://konghq.atlassian.net/browse/FT-2852

### Testing

[Preview](https://deploy-preview-4700--kongdocs.netlify.app/hub/kong-inc/app-dynamics/)

### CODE REVIEW

File: app/_hub/kong-inc/app-dynamics/_index.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+proxy requests handled by Kong can be identified and analyzed in
+AppDynamics.  The plugin reports request and response timestamps as
+well as error information to the AppDynamics platform so that they can
+be analyzed in the AppDynamics flow map and correlated with other
+systems participating in handling application API requests.
+
+The plugin utilizes the
+[AppDynamics C/C++ Application Agent and SDK](https://docs.appdynamics.com/pages/viewpage.action?pageId=42583435),
+which must be downloaded and installed on the machine or in the
+container that is running Kong Gateway.  Please refer to the
```
Comment (mrwanny) [inline app/_hub/kong-inc/app-dynamics/_index.md:42]: Does both control plane and data plane need the `AppDynamics C/C++ Application Agent and SDK` ?

File: app/_hub/kong-inc/app-dynamics/_index.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+AppDynamics SDK web page for platform support information.
+
+# Installation
+
+To use the AppDynamics plugin in Kong Gateway, the AppDynamics C/C++
+SDK must be installed on all nodes running Kong Gateway.  The SDK is
```
Comment (mrwanny) [inline app/_hub/kong-inc/app-dynamics/_index.md:48]: Both Control Planes and Data Planes?

File: app/_hub/kong-inc/app-dynamics/_index.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+library file.  It must be placed in one of the locations configured by
+the
+[system's shared library loader](https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html).
+Alternatively, the **LD_LIBRARY_PATH** environment variable can be set
+to the directory containing the **libappdynamics.so** file when
+starting Kong Gateway.
+
+If the AppDynamics plugin is enabled in the configuration, Kong
+Gateway will refuse to start if the **libappdynamics.so** file cannot
+be loaded.  The error message will be similar to this:
```
Comment (mrwanny) [inline app/_hub/kong-inc/app-dynamics/_index.md:62]: What if the plugin is enabled in a Hybrid installation, would kong data plane crash?

Comment (kikito) [inline app/_hub/kong-inc/app-dynamics/_index.md:62]: We need to include a section here for installing this in Konnect as well. @hbagdi do you know who should I involve in order to get this?

**Review by kikito — CHANGES_REQUESTED** [review]

File: app/_hub/kong-inc/app-dynamics/_index.md (hunk 3 above)
Comment (hanshuebner) [reply app/_hub/kong-inc/app-dynamics/_index.md:62]: Yes.  The plugin is required on all dataplane nodes.

File: app/_hub/kong-inc/app-dynamics/_index.md (hunk 2 above)
Comment (hanshuebner) [reply app/_hub/kong-inc/app-dynamics/_index.md:48]: I'm not entirely sure about the CP nodes, but I believe they need to have the plugin enabled in order to be able to access its schema.  When the plugin is enabled, the shared library must be present.

**Comment by cloudjumpercat** [comment]
@hanshuebner Thanks for submitting this PR! When you have the chance, could you change the base branch to `release/gateway-3.1`? Thanks!

**Comment by hanshuebner** [comment]
>  Thanks for submitting this PR! When you have the chance, could you change the base branch to release/gateway-3.1?

@cloudjumpercat Done!

**Comment by Guaris** [comment]
I edited this and I have some questions @hanshuebner I have questions about the following things in the environment variable table: 
- is the type field in the environment variables a data type? 
- Is `number` a data type for App Dynamics? I changed it to integer. I don't know `number` to be a valid data type in C++ and I couldn't find anything about it in their docs. Let me know if that needs to be handled differently. 
- `KONG_APPD_CONTROLLER_USE_SSL` is a boolean data type, should the default value be `true`? I left it as `on`. 
-  Is `secret_string` a data type? Or should that field be "string", and I can work the secret part into the description. 
- For `KONG_APPD_CONTROLLER_HTTP_PROXY_PORT`, is the default value `0` or should it be empty?

**Comment by hanshuebner** [comment]
Hi @Guaris 

> is the type field in the environment variables a data type?

Yes, in a loosely defined sense.

> Is number a data type for App Dynamics? I changed it to integer. I don't know number to be a valid data type in C++ and I couldn't find anything about it in the app dynamics docs. Let me know if that needs to be handled differently.

That works.

> KONG_APPD_CONTROLLER_USE_SSL is a boolean data type, should the default value be true? I left it as on.

`true`, `on` and `1` are interpreted as the true values, everything else is considered to be false.

> Is secret_string a data type? Or should that field be "string", and I can work the secret part into the description.

A `SECRET_STRING` is never shown in the logs and is a string otherwise.  It can also be specified as vault reference.  In that case, the vault reference is shown in the logs, but not the referenced value.

> For KONG_APPD_CONTROLLER_HTTP_PROXY_PORT, is the default value 0 or should it be empty?

Leaving it empty would make more sense.

Let me know if you need additional information.

-Hans

**Comment by Guaris** [comment]
@hanshuebner I made the changes. This looks good to me, we will wait on @kikito to approve this and i'll merge it in.

**Review by Guaris — APPROVED** [review]

File: app/_hub/kong-inc/app-dynamics/_index.md (hunk 3 above)
Comment (mikefero) [reply app/_hub/kong-inc/app-dynamics/_index.md]: @kikito This should have no impact on Konnect, but I would recommend describing a process that works with a Docker image. I mention no impact on Konnect as Koko is reliant on the data plane handling all the runtime logic and with the minimal configuration set available it is all being done through env.

I wouldn't want to hold the PR for a Docker process as this is something that could be done afterwards.

**Comment by Guaris** [comment]
@hanshuebner Can I merge this into the release branch or would you like more reviews?

**Comment by hanshuebner** [comment]
> Can I merge this into the release branch or would you like more reviews?

This can be merged, I think.

**Pull request merged**